### PR TITLE
Update README.adoc

### PR DESCRIPTION
**Pull request by johnwcowan**
I did some Englishing and rewrote the URI part of the spec to tighten constraints substantially.  I also allowed for the 1A to change in case of a different version of Unisig itself.

### CODE REVIEW

**Comment by lassik** [comment]
Thanks a lot for the quick turnaround!

Englishing very welcome and LGTM.

**Comment by lassik** [comment]
On tightening the requirements: I fully agree in spirit that things are better when everything is as uniform as possible. But the issue comes back to the definition of what a format is that I rambled about on the schemepersist list last year. A format is fundamentally a physical format (as opposed to an interpretation of its meaning). So the minimum viable reader algorithm essentially defines the bounds of the format.

In the case of Unisig, the reader reads seven bytes and gives up if those are not the right magic. It then reads the length byte, followed by UUID (16 arbitrary bytes) or URI (1..255 arbitrary bytes).

While a UUID and a URI may also be formats, in Unisig they are _inner_ formats nested within the outer format which only cares about how many bytes it reads. A typical Unisig implementation doesn't validate or parse the URI or UUID (beyond perhaps formatting the UUID in the canonical hex notation).

Since the outer format of Unisig allows arbitrary bytes, implementations would be divided into two camps: validating implementations that enforce correct URI/UUID syntax, and non-validating implementations that just read/write any bytes they are handed. A divergence would result between the two camps.

Since the outer format allows people to easily write code ignoring the rules of the inner formats, I think the spec (and more generally, all format specs) should be written to emphasize which parts are necessary features of the outer format according to the minimum viable reader algorithm. Parts that we intend to be the case ultimately have to be merely our recommendations.

That's why I wrote "recommended" all over the place. I know it sounds weird, but when writing I was constantly conscious of what we are able to enforce and what we are not. It sucks if people disregard our carefully thought out HTTP URIs, but ultimately the outer format does allow them to write anything they want in the URI bytes, as well as the UUID bytes.

TL;DR I think we should have a clear separation between format and validation. Format is the outermost physical format that people are forced to use. Validation is stuff that we want them to use but ultimately have no way of enforcing.

I think all formats are necessarily like this. Designers are just not typically conscious of the distinction (except as subconscious unease).

**Comment by lassik** [comment]
On the dates: I believe domain-year is a good enough disambiguator in the usual case. That's because our dates are variable-length strings, so another version of the same format published later in the same year can simply use domain-year-month, or domain-year-month-day, etc. So if people accidentally choose a too coarse-grained date representation, they can always remedy that by using finer-grained representations in new versions.

That was just my initial idea. You may have a deeper point behind always using YYYY-MM-DD that I missed.

**Comment by johnwcowan** [comment]
I understand your reasoning.  But I think that "must" is directed to the
creators of formats: they must create formats that look like this, and
other people can depend on it.  A generic validator should enforce this,
but a specific validator doesn't have to.

By the same token, "must" in R7RS is directed to the implementer of a
Scheme, analogous to the creator of a format.

As for the use of a date, here's the problem.  Organization A owns foo.com
in January 2020 and expects to continue to, so they use
foo.com/something#2020.  Then they lose the domain in February 2020 and it
is bought by unrelated organization B, who creates a format
foo.com/something#2020, not knowing that organization A has already done
so.  That is why we need granularity down to a day; this is independent of
the use of the date for versioning.

**Comment by lassik** [comment]
> As for the use of a date, here's the problem.  Organization A owns foo.com
in January 2020 and expects to continue to, so they use
foo.com/something#2020.  Then they lose the domain in February 2020 and it
is bought by unrelated organization B, who creates a format
foo.com/something#2020, not knowing that organization A has already done
so.  That is why we need granularity down to a day; this is independent of
the use of the date for versioning.

Ah, you are exactly right. Let's adopt your proposal.

The hour-minute-second thing is almost ridiculous, but I put it in there since the spec will hopefully be in active use for at least a couple decades. People have a way of finding unexpected fringe applications for long-lived tech, and ISO has a ready-made syntax for arbitrarily granular dates that happens to fit in URI fragments and HTML5 anchors as well. Not sure what to do about it.

The most likely use case for super-granular timestamps would be ad-hoc formats generated by some company's internal network servers when they talk to each other. In that case, perhaps they might as well use UUIDs.

**Comment by lassik** [comment]
The ad-hoc many-times-a-day formats could be something like a bunch of servers being updated with the newest Protocol Buffers -style definitions whenever a programmer pushes a commit.

**Comment by johnwcowan** [comment]
People who change schemas as often as that have worse problems.

Any comment on Unisig's own versioning?

**Comment by lassik** [comment]
A further problem with the dates is whether to mandate that they be put into fragments. Not all URL schemes have a fragment part. Web URLs are the nicest ones for sure, but they may no longer be the normal thing to use in a few decades. It's nice to leave the door open for other URI schemes.

Therefore, I would suggest that the date be allowed in any part of a URI. We can mandate that web URIs (http, https, and their future successors) use the fragment for the date.

**Comment by lassik** [comment]
> People who change schemas as often as that have worse problems.

They only have problems if schema changes are not orchestrated (orchestration is more difficult with each additional server; it's not terrible with only a couple of them and a tool like Ansible or Puppet).

I would still like to allow arbitrary timestamp granularity since ISO 8601 has done the design work for the syntax and it incurs no cost to people who want less granular timestamps.

**Comment by johnwcowan** [comment]
Okay then, anywhere in the URL, granularity a day or finer.

**Comment by lassik** [comment]
> Any comment on Unisig's own versioning?

I was hoping it could be unversioned like PNG and JSON, but your proposal to change the `^Z` to another byte is compelling. It's a bit of a philosophical debate whether the format is still the same if that byte is changed. Nevertheless, we can suggest that people who want an incompatible change can change that byte.

**Comment by johnwcowan** [comment]
I'd rather keep versioning in the hands of the Unisig Consortium, just like
Unicode.

**Comment by lassik** [comment]
> I'd rather keep versioning in the hands of the Unisig Consortium, just like Unicode.

Heh. Fair enough. Works for me.

**Comment by lassik** [comment]
> I would still like to allow arbitrary timestamp granularity since ISO 8601 has done the design work for the syntax and it incurs no cost to people who want less granular timestamps.

lol, probably every design committee eventually ends up sounding like this. "I would still like to let people breed black rabbits on the moon since we already defined how to breed rabbits and it incurs no cost to people who breed white rabbits on earth."

**Comment by lassik** [comment]
> I understand your reasoning.  But I think that "must" is directed to the creators of formats: they must create formats that look like this, and other people can depend on it.  A generic validator should enforce this, but a specific validator doesn't have to.
>
> By the same token, "must" in R7RS is directed to the implementer of a Scheme, analogous to the creator of a format.

That's a good point. I think we only disagree (if at all) about what terminology to use when talking about different levels of enforcement (or enforceability). I've never thought in depth about what terminology is appropriate, and simply improvised the language in the present document.
